Step3 - 로또(자동)

bestsilver opened the pull request:
안녕하세요!
step3 로또 자동 미션을 구현해보았습니다~!
이번주 메인 주제인 TDD로 구현해보기 위해 노력했지만 하다가 시간이 오래 걸려서 
완벽히 TDD로 구현하지 못했습니다...ㅠㅠ
익숙하지 않아서 그런지 많은 연습이 필요할 것 같네요ㅠㅠ
그리고 대부분 익셉션은 IllegalArgumentException 인데 따로 custom Exception을 만들어 빼야 할까요?
이번 미션에서는 custom으로 만들어 분리해보았습니다! 

시간 나실 때 리뷰 부탁드리겠습니다!
감사합니다!!

bestsilver committed:
docs: 기능 정의서 작성

bestsilver committed:
chore: empty파일 삭제

bestsilver committed:
test: Money 테스트 케이스 작성

- 숫자 이외의 값 테스트

bestsilver committed:
test: Lotto 테스트 케이스 작성

- 로또번호 범위 체크
- 로또번호 개수 체크
- 로또번호 중복 체크
- 로또번호_숫자_이외의_값_체크
- 로또_프린트_테스트

bestsilver committed:
test: Lottos 테스트 케이스 작성

- 로또들_프린트_테스트
- append_Lotto

bestsilver committed:
test: Figures 테스트 케이스 작성 (로또당첨통계)

- 당첨_통계_결과_테스트

bestsilver committed:
test: Profit 테스트 케이스 작성

- append_수익

bestsilver committed:
test: ProfitRate 테스트 케이스 작성

- 머니가_0인_경우_수익률
- 이익이_0인_경우_수익률
- 머니와_이익이_0인_경우_수익률
- 수익률_프린트_테스트

bestsilver committed:
feat: 로또 기능 구현

- 생성자 2개 사용
- 숫자가 아닌 문자가 왔을 경우 체크
- 로또 번호 범위가 아닌 숫자가 왔을 경우 체크
- 로또 번호 개수 체크
- 로또 번호 중복 체크
- 유효한 로또인지 체크

bestsilver committed:
test: LottoRank 테스트 케이스 작성

- 매치_카운트에_따른_랭크_얻기
- 로또_비교후_결과_랭크

bestsilver committed:
feat: 머니 기능 구현

- 스트링은 넘겨받아 int 형태로 변환
- 숫자가 아닌 경우 에러 throw

bestsilver committed:
feat: 로또들(로또리스트) 기능 구현

- 로또 append 기능
- 로또들을 출력하는 기능

bestsilver committed:
feat: 랭크 기능 구현

- enum으로 매치된 숫자개수, 당첨금에 따라 정의
- 매치된 숫자개수에 따라 랭크 리턴하는 기능
- 두개 로또 비교해서 매치된 숫자 개수에 따라 랭크를 리턴하는 기능

bestsilver committed:
feat: 당첨 통계 기능 구현

- 로또와 당첨 로또를 전달 받아 랭크에 따라 수익률 계산 및 통계화
- 당첨 통계를 출력하는 기능

bestsilver committed:
feat: 수익금 기능 구현

bestsilver committed:
feat: 수익률 기능 구현

- 수익금/원금
- 수익금 또는 원금이 0인지 체크
- 소수 2째짜리까지 프린트

bestsilver committed:
feat: 구매한 로또개수 구현

bestsilver committed:
feat: GameController 구현

- run에서 게임 시작됨
- 올바른 머니인지 체크
- 올바른 머니 리턴
- 올바른 로또인지 체크
- 올바른 로또 리턴
- 로또 당첨 통계 출력

bestsilver committed:
feat: UserInputView 구현

- 사용자 입력 기능 구현

bestsilver committed:
feat: Output 구현

- 게임 진행에 따른 안내글 출력
- 에러 출력

bestsilver committed:
feat: Exception 구현

- IllegalLottoException
- IllegalMoneyException

bestsilver committed:
feat: Application 구현

- 게임 실행

bestsilver committed:
Merge branch 'bestsilver' into step3

bestsilver committed:
refactor: Lottos append 위해서 바로 대입하지 않고 arrayList로 치환하여 넣어주기

bestsilver committed:
refactor: Arrays.asList 형태로 초기화 후 append 기능 테스트

bestsilver committed:
refactor: Arrays.asList 형태로 초기화 후 append 기능 테스트

bestsilver committed:
Merge remote-tracking branch 'upstream/bestsilver' into step3

bestsilver committed:
Merge remote-tracking branch 'upstream/bestsilver' into step3

bestsilver committed:
Merge remote-tracking branch 'upstream/bestsilver' into step3

bestsilver committed:
Merge remote-tracking branch 'upstream/bestsilver' into step3

bestsilver committed:
Merge remote-tracking branch 'upstream/bestsilver' into step3

bestsilver committed:
Merge remote-tracking branch 'upstream/step3' into step3

bestsilver committed:
Merge remote-tracking branch 'upstream/step3' into step3

bestsilver committed:
Merge remote-tracking branch 'upstream/step3' into step3

bestsilver commented:
어휴...깃에 익숙하지 않아서 삽질좀했네요ㅠㅠㅠ

bestsilver committed:
Merge remote-tracking branch 'upstream/bestsilver' into step3

bestsilver committed:
Merge remote-tracking branch 'upstream/bestsilver' into step3

bestsilver committed:
Merge remote-tracking branch 'upstream/bestsilver' into step3

bestsilver committed:
Merge remote-tracking branch 'upstream/step3' into step3

bestsilver committed:
style: STEP2와 STEP3 병합

enemfk777 commented on src/main/java/lotto_auto/exception/IllegalLottoException.java:
```diff
@@ -0,0 +1,14 @@
+    public static final IllegalLottoException NOT_NUMBER_ERROR = new IllegalLottoException(Lotto.NOT_NUMBER);
+    public static final IllegalLottoException NOT_MATCHED_NUMBER_SIZE_ERROR = new IllegalLottoException(Lotto.NOT_MATCHED_NUMBER_SIZE);
+    public static final IllegalLottoException EXIST_DUPLICATE_VALUE_ERROR = new IllegalLottoException(Lotto.EXIST_DUPLICATE_VALUE);
+    public static final IllegalLottoException NOT_RANGE_NUMBER_ERROR = new IllegalLottoException(Lotto.NOT_RANGE_NUMBER);
```
이전 step에서 부터 exception을 상수화 시키는 패턴을 사용하고 계시네요 :)
이유를 알 수 있을까요?
런타임에 오랜기간 동안 호출되지 않을 exception이 상수화 되어 있으면 어떤 문제가 있을까요? :)

enemfk777 commented on src/main/java/lotto_auto/controller/GameController.java:
```diff
@@ -0,0 +1,117 @@
+    private final Output output = new Output();
+
+    public GameController() {
+        this.money = getMoney();
```
Money를 Controller의 멤버변수로 설정해 두신 이유가 있을까요? :)
게임을 구매할 때 마다 새로운 컨트롤러가 만들어져야 할까요?

enemfk777 commented on src/main/java/lotto_auto/model/Money.java:
```diff
@@ -0,0 +1,20 @@
+
+import lotto_auto.exception.IllegalMoneyException;
+
+public class Money {
```
Money  객체를 생성할 때 -1000 원이 입력되면 어떻게 될까요? :)

enemfk777 commented on src/main/java/lotto_auto/model/Lotto.java:
```diff
@@ -0,0 +1,86 @@
+public class Lotto {
+    private final List<Integer> lottoNumbers;
+
+    public static final int validSize = 6;
+    public static final int minNumber = 1;
+    public static final int maxNumber = 45;
+    private static final String lottoDelimiter = ", ";
+    public static final int price = 1000;
```
상수의 변수명은 대문자 스네이크 케이스로 작성하는 게 컨벤션입니다 :)

enemfk777 commented on src/main/java/lotto_auto/model/PurchasedLottoCount.java:
```diff
@@ -0,0 +1,13 @@
+    int count;
+
+    public PurchasedLottoCount(Money money) {
+        this.count = money.getMoney()/Lotto.price;
```
PurchasedLottoCount 클래스가 갖는 역할과 기능은 무엇일까요?
view에 값을 전달하는 dto로 설계되었다기에는 입력 금액과 게임당 가격으로 로또 수량을 계산하는 도메인 로직이 있고
도메인 클래스라기에는 맡은 역할이 getter로 view에 값을 제공하는 일 밖에 없네요 :)

추가로
Lotto.price의 경우 어플리케이션 내에서 Money 클래스의 인스턴스가 의미를 갖기 위한 최소 단위라고 볼 수 있을 것 같은데요, 그렇다면  Lotto 클래스보단 Money 클래스에 위치하는게 더 응집도 높은 코드가 될 것 같습니다.
그렇게 된다면, Money 클래스에서 count를 스스로 반환할 수 있지 않을까요?

enemfk777 commented on src/main/java/lotto_auto/view/UserInputView.java:
```diff
@@ -0,0 +1,14 @@
+import java.util.Scanner;
+
+public class UserInputView {
+    private static final Scanner scan = new Scanner(System.in);
```
스캐너 캐싱 👍

enemfk777 commented on src/main/java/lotto_auto/controller/GameController.java:
```diff
@@ -0,0 +1,117 @@
+        List<Integer> numberList;
+
+        do {
+            Random generatorNumber = new Random();
+            IntStream numberStream =  generatorNumber.ints(Lotto.validSize, Lotto.minNumber, Lotto.maxNumber);
+            numberList = numberStream.boxed().collect(Collectors.toList());
+            isValid = isValidLotto(numberList);
```
랜덤 로또 번호 생성을 책임지는 클래스를 분리해보시는건 어떨까요?
또 로또 번호는 1 - 45 의 풀이 정해져 있는데, 캐싱된 숫자 풀을 사용해보시는건 어떨까요?
랜덤한 번호 6개를 생성할 땐 셔플을 해주면 될 것 같습니다 :) (Collections.shuffle 메서드 참조)

enemfk777 commented on src/main/java/lotto_auto/model/Lotto.java:
```diff
@@ -0,0 +1,86 @@
+import java.util.stream.Collectors;
+
+public class Lotto {
+    private final List<Integer> lottoNumbers;
```
Lotto라는 클래스는 숫자 6개로 이루어진 로또 한 게임의 역할과 기능을 갖는 클래스로 보여지는데요,
그렇다면 1 ~ 45 범위를 갖는 다는 제약사항 역시 Lotto의 책임으로 볼 수 있을까요?
멤버 변수로 Integer List를 갖기 때문에 Lotto가 해당 책임까지 떠안게 된 것으로 보이는데요,
"로또숫자"라는 클래스를 분리하고 Lotto는 "로또숫자" 클래스를 멤버 변수로 갖고,
숫자에 관한 제약사항은 "로또숫자" 클래스가 가져가는건 어떨까요? :)

enemfk777 commented on src/main/java/lotto_auto/model/LottoRank.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.Optional;
+
+public enum LottoRank {
+    FIRST(6, 2000000000),
```
2000000000 -> 2_000_000_000
긴 숫자는 위와 같이 1000 단위로 언더바를 넣어주면 가독성을 높일 수 있습니다 :)

enemfk777 commented on src/main/java/lotto_auto/model/LottoRank.java:
```diff
@@ -0,0 +1,49 @@
+        return rank.orElse(NOTHING);
+    }
+
+    public static LottoRank matches(Lotto o1, Lotto o2) {
```
LottoRank가 Lotto 까지 알 필요가 있을까요 :)?
LottoRank의 관심사는 "숫자 몇개를 맞췄나" 이고 역할은 그에 따른 Rank를 반환하는 것 일 것 같습니다.
로또 번호 몇개를 맞췄는지는 제공하는 역할은 LottoRank와 Lotto 중 어느 클래스가 더 어울릴까요??

enemfk777 commented on src/main/java/lotto_auto/model/Lottos.java:
```diff
@@ -0,0 +1,39 @@
+
+    private List<Lotto> lottoList;
+
+    public Lottos(List<Lotto> list) {
+        this.lottoList = new ArrayList<>();
+        this.lottoList.addAll(list);
+    }
```
테스트에서만 쓰이는 생성자 및 메서드는 사용을 지양하시는 편이 좋습니다.
테스트코드는 어디까지나 프로덕션 코드를 테스트하기 위해 존재하는 것이고,
테스트코드가 프로덕션 코드에 영향을 끼치는 것은 안티패턴 입니다 :)

enemfk777 commented on src/main/java/lotto_auto/model/Lottos.java:
```diff
@@ -0,0 +1,39 @@
+        this.lottoList.addAll(list);
+    }
+
+    public Lottos() {
+        lottoList = new ArrayList<>();
+    }
+
+    public void appendLotto(Lotto lotto) {
+        this.lottoList.add(lotto);
+    }
```
내부 컬렉션을 외부에서 조작하는 것은 지양하는 것이 좋습니다.
객체내부의 컬렉션 상태를 짐작하기 힘들게 하기 때문인데요,

이 로직은 삭제하고 테스트코드에서만 쓰이는 맨 위 생성자를 프로덕션코드에서도 사용하시는게 어떨까요?

enemfk777 commented on src/main/java/lotto_auto/view/Output.java:
```diff
@@ -0,0 +1,26 @@
+    public static final String lottoFiguresDelimiter = "----------";
+    public static final String profitRateGuideFormat = "총 수익률은 %s입니다.\n";
+
+    public void showNotice(String str) {
+        System.out.println(str);
+    }
+
+    public void showNotice(String strFormat, int value) {
+        System.out.format(strFormat, value);
+    }
+
+    public void showNotice(String strFormat, String value) {
+        System.out.format(strFormat, value);
+    }
```
같은 이름으로 인자만 다르게 메소드 오버로딩이 되어있어 코드를 읽는 입장에서 이게 어떤 목적으로 사용되었는지 파악하기가 쉽지 않습니다.

메세지를 인자로 받기보다는 각 메세지에 맞는 메서드를 분리하고, 그에 맞는 메서드명을 갖는 것이 코드를 읽는 입장에서 훨씬 이해하기 쉬울 것 같아요 :)

enemfk777 commented on src/main/java/lotto_auto/controller/GameController.java:
```diff
@@ -0,0 +1,117 @@
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class GameController {
```
전체적인 클래스 분리에 관한 내용이라 이곳에 적겠습니다 :)
도메인 클래스와 dto 클래스의 경계가 잘 나뉘어 지지 않은 것 같습니다

도메인 클래스 : 어플리케이션의 핵심 비즈니스 로직을 갖는 클래스 Lotto, Lottos, LottoRank 등
dto 클래스 : 값 전달을 위한 클래스 도메인 클래스가 view 클래스에 직접 노출되지 않게 view에 맞는 값을 전달

도메인클래스를 view에 직접 노출시키지 않는 것이 좋은 이유는 view에서 도메인클래스를 값 노출에 직접 사용하게 될 경우 view에서 보여주는 방식만 바뀌는 변화에도 도메인클래스가 직접 영향 받게 되기 때문입니다.

비즈니스의 핵심을 담고있는 도메인클래스가 자주 변경에 노출되면 그만큼 버그, 오류가 발생할 여지가 많아지게 되죠 :)
때문에 도메인클래스에서 핵심 로직을 다 수행하고 난 결과물만 dto 클래스를 통해 view로 전달하게 해서
만약 view에서 보여주는 값에 변화가 생겨도, dto 클래스까지만 영향 받게 하고, 도메인클래스를 보호하는 것입니다.

위 사항을 잘 고민하셔서 도메인 클래스와 dto 성격의 클래스가 잘 나뉘면 좋을 것 같습니다.
패키지로 나누어 보시는것도 좋을 것 같아요 :)

enemfk777 reviewed (changes requested):
안녕하세요 고은님 :)
주말에도 미션 진행하시느라 고생 많으셨습니다 🙇‍♂️
아직 객체간 책임 분리가 부족한 것 같아 몇가지 코멘트를 드렸습니다!
고민해보신 후 다시 한번 요청 부탁드립니다 !!

오늘도 화이팅입니다 💪

bestsilver committed:
docs: 추가 설계한 클래스들 기능 설명 추가

bestsilver committed:
refactor: 필요없는 에러 삭제

bestsilver committed:
refactor: Profit dto로 이동

bestsilver committed:
refactor: Profit Test 삭제

bestsilver committed:
refactor: 쓸대없는 클래스 삭제

bestsilver committed:
refactor: 매직넘버 삭제

bestsilver committed:
refactor:
- 상수 대문자 수정
- LottoNumbers 적용
- default rank count 상수화

bestsilver committed:
feat: LottoNumbers 기능 추가

클래스 책임 분리에 따라 LottoNumbers 추가
- 로또번호 개수 체크
- 로또번호 중복 체크
- 기본 생성자일 경우 랜덤으로 로또번호 생성

bestsilver committed:
feat: LottoNumber 기능 추가

클래스 책임 분리에 따라 LottoNumber 추가
- 로또번호 범위 체크
- 로또번호 숫자 아닌 값 체크

bestsilver committed:
refactor: 내부컬렉션 외부 조작 기능 삭제

bestsilver committed:
refactor: 로또끼리 비교기능 해당 위치로 이동

bestsilver committed:
refactor: 로또끼리 비교기능 삭제, 당첨금에 _넣어서 가독성 업

bestsilver committed:
refactor: 머니 최소 사이즈 예외 추가, lotto price 해당 위치로 옮기기, 살수있는 로또 개수 구하기 기능 추가

bestsilver committed:
refactor: notice 기능 분리

bestsilver committed:
refactor: 설계 변경에 따른 기능 호출 변경

bestsilver commented:
클래스 분리를 계속 해보았는데 너무 어렵네요..ㅠㅠ
아직도 헷갈리는 부분이 많아서 많은 연습이 필요할 것 같아요...ㅠㅠㅠㅠ

bestsilver committed:
refactor: 구매한 로또 개수 출력 추가

enemfk777 commented on src/main/java/lotto_auto/model/Money.java:
```diff
@@ -1,20 +1,36 @@
+    public int canBuyLottoCount() {
+        if (money < LOTTO_PRICE) {
+            return 0;
+        }
```
이부부도 객체 생성시 검증 하면 어떨까요?
1000 원이하면 Exception을 던지면 될것 같아요
Money가 어플리케이션에서 의미를 갖으려면 1000원 이상이어야 할 테니까요 :)

enemfk777 commented on src/main/java/lotto_auto/model/LottoNumbers.java:
```diff
@@ -0,0 +1,70 @@
+                .collect(Collectors.toList());
+        Collections.shuffle(numbers);
+        return numbers.subList(0, LottoNumbers.VALID_SIZE);
+    }
```
개발자가 제어할 수 없는 코드가 포함되어 있어 테스트가 불가능한 클래스가 되었습니다.
Instant.now(), Random 등은 실행 할때마다 결과가 다른, 즉 개발자가 제어할 수 없는 코드들 인데요

로또 미션의 TDD로 자동차 경주 게임 구현 아티클 (https://edu.nextstep.camp/s/X02BsEA0/ls/PBl6en4H)을 읽어 보신후 
테스트하기 어려운 부분을 찾아 테스트 가능한 구조로 개선
을 적용해 보시면 어떨까요?

enemfk777 commented on src/main/java/lotto_auto/model/LottoNumber.java:
```diff
@@ -0,0 +1,51 @@
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException(NOT_NUMBER);
+        }
+    }
```
굳이 생성자가 두개 일 필요는 없어 보입니다. 하나로 통일 해도 괜찮을 것 같아요 :)

enemfk777 commented on src/main/java/lotto_auto/model/LottoNumbers.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    public LottoNumbers(String stringNumbers) {
+        this.lottoNumberList = Arrays.stream(stringNumbers.split(LOTTO_DELIMITER))
```
문자열을 구분자로 나누는 책임이 LottoNumbers의 책임일까요 ? :)

enemfk777 commented on src/main/java/lotto_auto/model/LottoNumbers.java:
```diff
@@ -0,0 +1,70 @@
+import java.util.stream.IntStream;
+
+public class LottoNumbers {
+    List<LottoNumber> lottoNumberList;
```
중복되지 않은 숫자라는 요구사항을 충족하기 위한 자료구조는 List 외에 어떤 것이 있을까요? :)

enemfk777 commented on src/main/java/lotto_auto/dto/Profit.java:
```diff
@@ -0,0 +1,13 @@
+package lotto_auto.dto;
+
+public class Profit {
+    int profit;
+
+    public Profit(int profit) {
+        this.profit = profit;
+    }
+
+    public int getProfit() {
+        return profit;
+    }
+}
```
ProfitRate에게 기능을 빼앗겨 존재 의미가 없는 클래스가 되어벼렸네요 :)

enemfk777 commented on src/main/java/lotto_auto/model/Figures.java:
```diff
@@ -0,0 +1,45 @@
+        return winningMoney;
+    }
+
+    public String printFigures() {
+        StringBuilder result = new StringBuilder();
+        for (LottoRank rank : LottoRank.valuesTheLowestOrder()) {
+            result.append(String.format(FIGURES_FORMAT, rank.matchedCount(), rank.winnings(), figures.get(rank)));
+        }
+
+        return result.toString();
+    }
```
출력 책임은 View에서 지는것이자연스러울것 같아요
Figures와 같은 도메인클래스는 출력에 필요한 값만 반환하도록 하구요 :)

현재 도메인 클래스와 View가 모두 Java로 되어있어 더욱 헷갈리시 것 같은데

실무에서 API와 API 값을 출력하는 프론트엔드에 투영해보시면 이해가 쉬우실 것 같아요 :)

enemfk777 reviewed (changes requested):
안녕하세요 고은님 :)
일요일에도 미션 진행하시느라 고생 많으셨습니다 🙇‍♂️
다음 단계로 가기전에 쪼오금 더 발전 시킬 수 있는 부분이 보여서 한 번 더 Request Changes 드립니다!!
로또 TDD 미션의 [TDD로 자동차 경주 게임 구현 문서](https://edu.nextstep.camp/s/X02BsEA0/ls/PBl6en4H)를 읽어보시고 테스트 가능한 구조로 개선하는 방법에 대해 고민해 보시면 좋을 것 같습니다.

그럼 화이팅입니다 💪

bestsilver committed:
test: 수익금 결과 테스트 추가

bestsilver committed:
test:
- 로또번호 숫자 이외의 값 테스트 이동
- LottoNumbers 외부에서 add시 에러 발생 테스트 추가

bestsilver committed:
test: 외부에서 add시 에러 발생 테스트 추가

bestsilver committed:
test: ParameterizedTest 형식으로 변경

bestsilver committed:
test: 기준 값 이하 테스트로 변경

bestsilver committed:
test: 머니는 0인 경우 에러 발생, 이익이 0인 경우 위주로 테스트

bestsilver committed:
test: count에 맞는 로또수 발행되는 지 테스트 추가

bestsilver committed:
refactor: 프린트 로직 이동, figures를 넘겨주지 않고 rank에 따라 count 넘겨주는 방식으로 변경

bestsilver committed:
refactor: 프린트 로직 이동

bestsilver committed:
refactor: 필요없는 생성자 삭제

bestsilver committed:
refactor: List를 Set으로 변경하여 Duplicate 로직 삭제

bestsilver committed:
refactor: lottoList unmodifiable로 변경

bestsilver committed:
refactor: 누락된 접근 제어자 추가, money 최소값 변경

bestsilver committed:
refactor: 사용하지 않는 Profit Class 삭제

bestsilver committed:
feat: LotttoGenerator Class 추가, 로또 및 로또들 생성

bestsilver committed:
refactor: buyLottos 로직 lottoGenerator 사용하도록 수정

bestsilver committed:
refactor: LottoNumbers에서 string argument 생성자 삭제로 인한 변경

bestsilver committed:
refactor: userinput delimiter로 자르는 로직 controller로 이동

bestsilver committed:
refactor: string argument 생성자 삭제

bestsilver committed:
refactor: 숫자 이외 문자 에러 message 문자 numbers에서 number로 이동

bestsilver commented:
안녕하세요! 주말 잘 지내셨나요!!
리뷰하시느라 고생 많으시네요 ㅠㅠ

이번에는  테스트가 불가능한 클래스에 대해 수정 및 필요없는 클래스 삭제, List -> Set 변경 등
리뷰 해주신 부분에 대해 수정 및 테스트 코드들 약간 추가해보았습니다!

enemfk777 commented on src/main/java/lotto_auto/model/LottoGenerator.java:
```diff
@@ -0,0 +1,35 @@
+    public static Lottos createLottos(int count) {
+        List<Lotto> lottos = new ArrayList<>();
+        for (int i=0; i < count; i++) {
+            lottos.add(createLotto());
```
메서드 내에서 클래스 내부의 다른 퍼블릭 메서드를 호출하는 것은 지양하는 편이 좋습니다 :)
퍼블릭 메서드는 협력하는 다른 객체와 메세지를 주고받는 용도로 존재하는 메서드이기 때문에
내부의 다른 메서드에서 사용하는 순간 객체지향 프로그래밍 원칙 중 단일 책임 원칙을 어기는 것으로 볼 수 있을 것 같습니다.

이 원칙을 어김으로써 아래와 같은 문제가 일어날 것 같은데요

> A라는 객체와 협력하기 위해 만든 B라는 메서드를
> C라는 객체와 협력하기위해 존재하는 D라는 메서드 내부에서 사용하게되면,
> A라는 객체와 협력하는데 필요한 요구사항이 바뀌어 B를 재정의 해야할 때
> A와 관계없는 D까지 영향을 받게 되어버립니다.

createLotto가 createLottos 내부에서만 쓰인다면 접근제한자를 private으로 두는건 어떨까요?

enemfk777 commented on src/main/java/lotto_auto/model/LottoGenerator.java:
```diff
@@ -0,0 +1,35 @@
+import static lotto_auto.model.LottoNumbers.VALID_SIZE;
+
+public class LottoGenerator {
+    public static final List<Integer> LOTTO_ALL_NUMBERS = IntStream
+            .rangeClosed(MIN_NUMBER, MAX_NUMBER)
+            .boxed()
+            .collect(Collectors.toList());
```
스트림 API 사용으로 숫자 풀 캐싱 👍
Integer가 아니라 LottoNumber 타입의 풀을 갖는건 어떨까요 :)

enemfk777 reviewed (approved):
안녕하세요 고은님 :)
코멘트 드렸던 사항을 잘 반영해주셨네요 💯
소소한 코멘트를 추가로 드렸으니 다음 단계에서 함께 반영 부탁드려요!! :)
이번 단계는 이만 머지하겠습니다! 고생하셨습니다 !

다음 단계도 화이팅입니다 💪
